### DISCUSSION TITLE
如何在编译时为每个源文件添加单独的宏

### DISCUSSION THREAD

**Discussion by shenxing1206** [xmake-io/xmake#3392] [Q&A] [answered]
场景是在日志输出时需要打印当前文件名，不带目录。

在makefile中可以直接以如下方式写这个宏
`%.o: %.cpp

    $(CC) $(headers) $(CFLAGS) -o $@ -c $^ -D__FILEFILE__="$(notdir $<)"
`

请问在xmake中要如何实现此功能呢？
猜测是类似于：
    before_build_file(function (target, sourcefile, opt)
        add_defines(format("__FILENAME__=%s", path.filename(sourcefile)))
    end)
但我不知道具体的语句应该怎么写？能否给一个正确的示例。
或者是否有更简单的方式？

**Comment by waruqi** [comment, marked as answer]
直接代码里走 `__FILE__` ，解析提取下名字不就完事了，干嘛这么麻烦，都回显 logs 了，也不差这点性能。

```lua
target("")
    for _, file in ipairs(os.files("src/*.c")) do
        add_files(file, {defines = "FILENAME=" .. path.filename(file)})
    end
```
